Saisiddu/merge runon to v1

saisiddu opened the pull request:
This pull request addresses the following:
- updating the components Radiation, PotentialEvapotranpiration, SoilMoisture, Vegetation, and VegCA. 
- committed new components ResourceRedistribution and SpatialDisturbance.

Once the pull request is merged, I will request for the creation of a new release that will be included in the Landlab ecohydrology paper.

saisiddu commented:
@mcflugen @kbarnhart or anybody: It looks like 'test_corners_at_cell' and 'test_faces_at_cell' @ C:\projects\landlab\landlab\graph\radial\tests\test_dual_radial.py are failing (from the report from appveyor on Py3_7). I updated my branch from v1, edited ecohydro components, and have pulled from v1 again before creating the pull_request. Did I mess any of these files? Can you please help me address these failures? Thank you!

mcflugen commented:
@saisiddu We probably no longer need to support Python 2.7, even on landlab v1, so I think it would be fine if you just removed those builds.

The other issues look to be fairly minor. There seems to be an issue with what the starting element is when ordering elements in a rotational sense. I don't think it's a big issue, particularly because this is on the *v1* branch. These errors appear on the current version of the *v1* branch so it's not something that you've broken.

saisiddu commented:
@mcflugen Thank you! In that case, I would like to merge once someone approves the changes.

kbarnhart commented:
@saisiddu I have a couple of concerns about documentation and tests. 

I am not able to undertake a thorough review now, so I think it would be best if we found someone else who could do this. I should be able to do this by the end of this week if necessary. However, there are a few things that I noticed while doing a quick look through the code that I wanted to raise. 

1) Not all the component inputs are documented in the docstring (e.g., in the PotentialEvapotranspiration and SoilMoisture component init). I think its really important that these are all documented. 
2) There are not that many tests, and its unclear to me whether tests ensuring component behavior exist. Perhaps I am mistaken, but I suspect a few more are needed. I saw that the docstrings have some examples that serve as tests, and that there are unit tests, but they are primarily of the interface (e.g., ensuring that component metadata exists). I raise this point especially because you have a couple of components which have flags indicating different methods. I think that the tests should cover those different methods. Since coverage isn't building on these (since the didn't succeed), its hard to tell how well covered these components, are. I might consider running coverage locally to test.
3) Both the ResourceRedistribution and SoilMoisture components have examples, but it was really hard for me to follow what the examples were of doing.

mcflugen commented:
@saisiddu I would rather not merge anything without all of the tests passing, even if it is into to the *v1* branch.

As part of this pull request, could you:
- [ ] remove the Python 2.7 builds
- [ ] fix the failing doctests (I think all that needs to be done is to change the expected output, but I can help with that)

Once the tests are passing, it will be easier to review.

saisiddu commented:
Thanks, @kbarnhart!
1) I will look for the overlooked inputs in PotentialEvapotranspiration & SoilMoisture and add them.
2) I will add tests for components that have different methods (Especially PotentialEvapotranspiration & SoilMoisture). That is a good catch. For most other cases, I added doctests that actually test if the component works. I agree that they can be improved but I might have to work on them at a later time.
3) I can understand why the examples might have been difficult to follow. I added extensive examples for the ResourceRedistribution component. However, I expect the user to refer the publication to actually learn about the processes implemented in the component. I don't intend someone to completely perceive the functioning of the CA model through my component's documentation. I might be able to add extensive documentation after my dissertation is submitted.

saisiddu commented:
Thanks, @mcflugen.

- I tried to fix the failing tests in test_dual_radial.py. I don't understand why the assertions failed. I ended up getting the output listed in the tests. Can you please look at them too?

saisiddu commented:
@mcflugen, @kbarnhart: I uninstalled and reinstalled Landlab in the landlab-dev enviornment. Tests are running on my machine (PY3.7). I am not sure what is the issue here. Can one of you check? How do we run these tests again on appveyor and travis-ci?

saisiddu commented:
@mcflugen : Travis builds failed at the linting step (it says pip couldn't install flake8). Is there something I could do to fix that? Can we skip that step so that docs can be tested?

@kbarnhart:
I added doctests to cover alternative methods (in SoilMoisture and PotentialEvapotranspiration). Local coverage shows that all ecohydro components have a mean coverage of ~90% except SoilMoisture. I suspect that my tests don't cover all 'if then else' blocks.

I would like to merge at this point and add more tests when the paper is in review.

mcflugen commented:
@saisiddu You'll need to fix all of the errors reported in the linting stage first. If you run `make lint`, you can run this step locally. Running the `black` command (e.g. `black landlab/components/pet/potential_evapotranspiration_field.py`) on the files will probably get you most of the way to eliminating the errors.

saisiddu commented:
Thanks @mcflugen.

saisiddu commented:
@mcflugen: Do I have to install some package to run lint locally? I am on windows10.

mcflugen commented:
@saisiddu Yes, you need to install *flake8* to run the linter and *black* for the reformatting. The following will install both,
```bash
$ pip install flake8 black
```

saisiddu commented:
@mcflugen: I was able to run 'black' command and flake8 is already installed. But I wasn't able to run 'make lint'. I couldn't find anything online yet. I tried running 'make lint' from both landlab and landlab\landlab folders.

mcflugen commented:
@saisiddu Try running the following from the top-level folder (the one with *setup.py*),
```bash
$ flake8 landlab --exclude=examples
```

saisiddu commented:
@mcflugen: Thanks, Eric. I ran flake8 for each of the ecohdyro components.

I will run the command you suggested from top folder now.

kbarnhart commented:
@saisiddu nice work improving the documentation and tests. 🎉 

1. A minor comment: I think you need to remove all of these files under .idea/
2. I think that it is worth having someone besides me do a review (not entirely certain if @mcflugen has looked at the code thoroughly, or was waiting for things to pass). I've added getting someone else to commit to doing a review soon to our call agenda for today. 

A side comment for @mcflugen: when I was working with @DavidLitwin he also had issues with make lint and make pretty from the command line (on a PC). I suspect its because make doesn't work out of the getgo on PC (this is speculation). Intended here as a note that we shouldn't expect those instructions to work for PC...

DavidLitwin commented:
@kbarnhart I should add that the resolution for PC was to use `black landlab`, but when I tried this recently, it wanted to change additional things that seem to pass `make lint`. Not sure why this is.

mcflugen commented:
@kbarnhart, @saisiddu, @DavidLitwin The problem with `make lint`, `make pretty`, etc. not working on Windows is that the *make* is a Unix command. I'm sure there's some way to install it on Windows but it's probably easier to just run the underlying commands (i.e. `flake8`, `black`, `isort`, etc.).

kbarnhart commented:
We just discussed this on the call and @mcflugen noted that 
a. he was waiting for the test to pass to review and
b. he will review this by the end of the week.

saisiddu commented:
Thanks, @kbarnhart, and @mcflugen.

I will remove those files under .idea (pycharm files).

mcflugen commented on landlab/components/plant_competition_ca/plant_competition_ca.py:
```diff
@@ -315,31 +347,44 @@ def update(self, time_elapsed=1, Edit_VegCov=True):
             ] = 1
 
         # For debugging purposes
+#        self._n = n
```
Remove commented-out code

saisiddu replied:
done

saisiddu replied:
done

mcflugen commented on landlab/components/resource_redistribution/resource_redistribution.py:
```diff
@@ -0,0 +1,738 @@
+
+        name = "vegetation__plant_functional_type"
+        if name not in self.grid.at_cell:
+            print(
```
Use *warning.warn* or, if you don't like that, print to *stderr*.

saisiddu replied:
done

mcflugen commented on landlab/components/resource_redistribution/resource_redistribution.py:
```diff
@@ -0,0 +1,738 @@
+            **kwds,
+        )
+
+    def initialize(
```
This method seems unnecessary? Why not just initialize these variables in *__init__*?

saisiddu replied:
This is intentional. The reason I have another method is that I would like to keep the option of re-initializing the inputs (probably a subset of them) and use the same object to do so.

mcflugen commented on landlab/components/resource_redistribution/tests/test_resource_distribution.py:
```diff
@@ -0,0 +1,78 @@
+"""
```
Could you please add more extensive tests that better test the functioning of the component?

saisiddu replied:
done

mcflugen commented on landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py:
```diff
@@ -0,0 +1,79 @@
+"""
+
+import numpy as np
+from landlab.components import FlowRouter  # , DepressionFinderAndRouter
```
Remove commented-out code.

saisiddu replied:
done

mcflugen commented on landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py:
```diff
@@ -0,0 +1,79 @@
+    channel length (upstream cell order).
+
+    Parameters:
+    ==========
```
Use *-* for underline, not *=*. Remove the colon.

saisiddu replied:
done

mcflugen commented on landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py:
```diff
@@ -0,0 +1,79 @@
+    outlet_id: int (Optional)
+        Outlet id to be set
+
+    Returns:
```
Same as above.

saisiddu replied:
done

mcflugen commented on landlab/components/soil_moisture/funcs_for_including_runon_in_soil_moisture.py:
```diff
@@ -0,0 +1,79 @@
+
+    if outlet_id is None:
+        outlet_id = np.argmin(grid.at_node['topographic__elevation'])
+    _ = grid.set_watershed_boundary_condition_outlet_id(
```
Remove the underscore.

saisiddu replied:
done

mcflugen commented on landlab/components/spatial_disturbance/tests/test_spatial_disturbance.py:
```diff
@@ -0,0 +1,75 @@
+"""
```
Could you please add better tests of the component?

saisiddu replied:
Can you please give me more pointers? What do you mean by better tests?

saisiddu replied:
done

mcflugen commented on landlab/components/spatial_disturbance/spatial_disturbance.py:
```diff
@@ -0,0 +1,527 @@
+        """
+        self._pft_scheme = pft_scheme
+        _assert_pft_scheme_is_valid(self._pft_scheme)
+        #        super(SpatialDisturbance, self).__init__(grid, **kwds)
```
Remove commented-out code.

saisiddu replied:
done

saisiddu commented:
@mcflugen:
- I prefer not changing the case of the variables for this version. This can be taken up for version 2. I have a lot of drivers (that have contributed to the results for the paper) that are using this interface and I don't think I can deal with the changes that might be required.

- Which notebooks are you referring to? Tutorials? There are tutorials but I don't think they need to be edited for these commits (as the interface hasn't changed). I will have to add new tutorials for the new components (ResourceRedistribution and SpatialDisturbace) but it will have to wait.

- Do you suggest I move some doctests to the unit tests?

mcflugen commented:
@saisiddu 
* I was just wondering if you had written any Jupyter Notebooks for the new components. If so, we would have to move them into this repository (from, maybe, the tutorials repository) when we merge all of this into landlab 2.
* Don't forget to remove the `.idea` folder.
* You may want to move some parts of your *doctests* into unit tests. It seems like the unit tests for the new components don't really test if the components are working properly but, rather, just general *landlab* component tests (e.g. if the component has a name). When we move these components to landlab 2 it will be important to make sure things are still working and that we didn't break anything in the move. As such, lots of good tests will be extremely helpful.

saisiddu commented:
Thanks, @mcflugen! I have addressed your comments, added unit tests for the new components (that test the entire component). I did remove `.idea` folder. Is it still around? 

@mcflugen, @kbarnhart : I am not sure why one of the Travis builds is failing while the other passed. Can you please look at this?

mcflugen commented:
@saisiddu looks good. It appears the *.idea* folder is still there, though. Could you remove that?

saisiddu commented:
Thanks a lot, @mcflugen! Merging this.
